### PR TITLE
repo: unify package repository representations

### PR DESCRIPTION
**Pull request by tigarmo**
This commit unifies the repo models from `package_repositories` and `projects`. Remarks:

- The Apt, AptDeb, AptPPA and AptUCA classes are removed completely from the `projects` module. The verification that they performed is now done by their "duals" in `package_repositories`.
- `projects.validate_repository()` is preserved as it's part of the "official" public API. Its implementation is changed to just call PackageRepository.unmarshal(), with (hopefully) no change in semantics.
- The `PackageRepositoryApt.name` property is removed; as far as I can tell only test code used it and it's not documented (it looks like it's not meant to be used by client code), so might as well remove it while we're cleaning things up.
- The majority of the model validation now generates Pydantic errors instead of the custom `errors.PackageRepositoryValidationError`. This is consistent with what the applications do but necessitates a small loss in "error clarity" because we remove the "resolution" and "details" fields. Hopefully the messages are still sufficiently clear.
- There were many duplicated tests between `test_package_repository.py` and `test_projects.py`, because each set performed the same validation tests in the two model classes "types". Now the majority of the tests are in `test_package_repository.py`.
- A minor point; the "priority" field is read as either an integer or one of the valid string values, but stored internally always as an integer (or None).

- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----

CRAFT-1665

### CODE REVIEW

**Comment by tigarmo** [comment]
Apologies in advance for how annoying this is going to be to review, heh
Tested these changes in snapcraft and rockcraft; Snapcraft required [minor changes](https://github.com/snapcore/snapcraft/pull/4107/files), things look otherwise fine.

**Comment by tigarmo** [comment]
Some of the now-uncovered lines in apt_sources_manager are from checks that are now impossible to hit, because the Pydantic validation already handles it:
image
I think we should just remove those codepaths

File: tests/unit/repo/test_package_repository.py [hunk 1]
```diff
@@ -64,160 +64,222 @@ def apt_repository():
...
+    ],
+)
+def test_apt_key_id(key_id, error):
+    repo = {
+        "type": "apt",
+        "url": "https://some/url",
+        "key-id": key_id,
+    }
+
+    if not error:
```
Comment (lengau) [inline tests/unit/repo/test_package_repository.py:250]: It's probably best to break this into two tests — one that tests success and one that tests failures

**Review by lengau — COMMENTED** [review]
This is **glorious!**

There's still some commented out code in the tests, so let me know when you're ready for me to re-review that. The comments are very minor.

File: craft_archives/repo/package_repository.py [hunk 2]
```diff
@@ -33,6 +34,13 @@
 UCA_DEFAULT_POCKET = UCA_VALID_POCKETS[0]
 UCA_KEY_ID = "391A9AA2147192839E9DB0315EDB1B62EC4926EA"
 
+# Workaround for mypy
+# see https://github.com/samuelcolvin/pydantic/issues/975#issuecomment-551147305
+if TYPE_CHECKING:
+    KeyIdStr = str
+else:
+    KeyIdStr = constr(regex=r"^[0-9A-F]{40}$")
```
[the remark this answers is not in the document]
Comment (tigarmo) [reply craft_archives/repo/package_repository.py]: Nice! I applied it "manually", thanks

File: craft_archives/repo/package_repository.py [hunk 3]
```diff
@@ -372,393 +209,105 @@ def pin(self) -> str:
...
-            data["key-server"] = self.key_server
-
-        data["name"] = self.name
-
-        if self.path:
-            data["path"] = self.path
-
-        if self.suites:
-            data["suites"] = self.suites
+    url: str
```
[the remark this answers is not in the document]
Comment (tigarmo) [reply craft_archives/repo/package_repository.py]: Yes! Updated, and also learned about [error_msg_templates](https://docs.pydantic.dev/usage/model_config/#options) to preserve the error message yay

File: tests/unit/repo/test_package_repository.py (hunk 1 above)
Comment (tigarmo) [reply tests/unit/repo/test_package_repository.py]: You are right of course! Updated

**Review by lengau — APPROVED** [review]
This all looks great! Thanks!

File: craft_archives/repo/package_repository.py [hunk 4]
```diff
@@ -45,15 +54,64 @@ class PriorityString(enum.IntEnum):
...
+        # pyright: reportUnknownVariableType=false
+        # pyright: reportUnknownLambdaType=false
+
+        validate_assignment = True
+        allow_mutation = False
+        allow_population_by_field_name = True
+        alias_generator = lambda s: s.replace("_", "-")  # noqa: E731
+        extra = "forbid"
+
+    type: Literal["apt"]
```
Comment (lengau) [inline craft_archives/repo/package_repository.py:73]: Not something we have to address, but I glanced at the changes you made for snapcraft, and defaulting this value to "apt" here (or as overrides in the two concrete classes) would ensure complete backwards compatibility if we want that.

Comment (tigarmo) [reply craft_archives/repo/package_repository.py:73]: I thought about doing that, but wouldn't it mean that users (the ones writing the `snapcraft.yaml` or `rockcraft.yaml`) would be able to declare package repositories _without_ specifying the type?

**Comment by tigarmo** [comment]
Running Snapcraft spread again with the latest changes just for sanity's sake

**Pull request merged**